### PR TITLE
jgm/daos-4698-1: Rebuild object migration should execute applicable object punch for each migration.

### PR DESCRIPTION
**Pull request by jgmoore-or**
Added object punch epoch to object enumeration.  Execute the punch during migrate.

### CODE REVIEW

File: src/object/obj_enum.c [hunk 1]
```diff
@@ -262,6 +265,22 @@ fill_key(daos_handle_t ih, vos_iter_entry_t *key_ent, struct dss_enum_arg *arg,
...
+		arg->kds_len++;
+
+		D_ASSERT(iov->iov_len + pi_size < iov->iov_buf_len);
+		memcpy(iov->iov_buf + iov->iov_len, &key_ent->ie_obj_punch,
+		       pi_size);
+
+		iov->iov_len += pi_size;
+		arg->obj_punched = true;
+	}
+
```
Comment (wangdi1) [inline src/object/obj_enum.c:283]: I thought we will get object punch epoch by scan. see rebuild_obj_scan_cb() L601.  So you do not need change enumerate packing.  
Anything concern?

Comment (jgmoore-or) [reply src/object/obj_enum.c:283]: I discussed this option with Jeff.  He says the scan only picks up the latest punch.  For the incarnation log problem we're trying to solve, the subsequent punch closest to the update epoch must me replayed.  There can be multiple object punches at differing epochs for an object, and recording them in the enumeration packing is required.

Comment (jolivier23) [reply src/object/obj_enum.c:283]: The latest isn't sufficient.  You could have many snapshots you are rebuilding.   As such, you could have something like this

update object at time 1
take snapshot at 2
punch at 3
update a at 4
snapshot at 5
punch at 6

If you rebuild the object at 7, you will only see the punch at 6 in the scan phase.   You only see the intermediate punch 3 while you are doing migration for the snapshot at 2.   So you can't do this at the scan phase.

**Review by jolivier23 — APPROVED** [review]

File: src/object/obj_enum.c (hunk 1 above)
Comment (wangdi1) [reply src/object/obj_enum.c:283]: ah, right. Thanks for explanation.

File: src/object/srv_obj_migrate.c [hunk 2]
```diff
@@ -1326,6 +1327,19 @@ migrate_dkey(struct migrate_pool_tls *tls, struct migrate_one *mrone)
 		D_GOTO(obj_close, rc);
 	}
 
+	/* punch the object */
+	if (mrone->mo_obj_punch_eph) {
+		rc = vos_obj_punch(cont->sc_hdl, mrone->mo_oid,
+				   mrone->mo_obj_punch_eph,
+				   tls->mpt_version, VOS_OF_REPLAY_PC,
+				   NULL, 0, NULL, NULL);
```
Comment (wangdi1) [inline src/object/srv_obj_migrate.c:1335]: @jolivier23  is it ok for VOS to punch the same object with same epoch multiple times?

File: src/object/srv_obj_migrate.c [hunk 3]
```diff
@@ -1326,6 +1327,19 @@ migrate_dkey(struct migrate_pool_tls *tls, struct migrate_one *mrone)
...
+		if (rc) {
+			D_ERROR(DF_UOID" punch obj failed: rc %d\n",
+				DP_UOID(mrone->mo_oid), rc);
+			return rc;
```
Comment (wangdi1) [inline src/object/srv_obj_migrate.c:1339]: D_GOTO(obj_close, rc) ?

**Review by wangdi1 — COMMENTED** [review]
Probably add a few more snapshots and punch in rebuild_snap_punch_empty() test. Thanks.

File: src/object/srv_obj_migrate.c (hunk 3 above)
Comment (jgmoore-or) [reply src/object/srv_obj_migrate.c]: This has been fixed.

File: src/object/srv_obj_migrate.c (hunk 2 above)
Comment (jolivier23) [reply src/object/srv_obj_migrate.c:1335]: It should be ok with the VOS_OF_REPLAY_PC flag.  If it doesn't work, it's a bug

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Need to revert the submodule update

**Comment by johannlombardi** [comment]
Should be ported to 1.2 since this is a 1.2 blocker.

**Pull request merged**